### PR TITLE
Fix 10193: Compiler crash with decorator and two "export default"s

### LINKED ISSUE #10193: Compiler crash with decorator and two "export default"s
**TypeScript Version:**  nightly (2.1.0-dev.20160806)

**Code**

Here's package.json to show the dependencies for this example:

```json
{
  "name": "compiler-bug",
  "version": "1.0.0",
  "description": "",
  "main": "test.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "mobx": "^2.4.2",
    "mobx-react": "^3.5.4",
    "react": "^15.3.0",
    "react-dom": "^15.3.0"
  }
}
```

Here is `test.tsx`:

```ts
import {observer} from "mobx-react";

@observer
export default class Foo {
	render() {
		return <div>Hi</div>;
	}
}

export default {
	uhoh: "another default",
};
```

**Expected behavior:**

An error for repeated `export default`

**Actual behavior:** 

`tsc --jsx react --experimentalDecorators test.tsx`

The following exception:

```
/usr/local/lib/node_modules/typescript/lib/tsc.js:37458
                throw e;
                ^

TypeError: Cannot read property 'kind' of undefined
    at getTypeForVariableLikeDeclaration (/usr/local/lib/node_modules/typescript/lib/tsc.js:15468:42)
    at getWidenedTypeForVariableLikeDeclaration (/usr/local/lib/node_modules/typescript/lib/tsc.js:15570:24)
    at getTypeOfVariableOrParameterOrProperty (/usr/local/lib/node_modules/typescript/lib/tsc.js:15621:28)
    at getTypeOfSymbol (/usr/local/lib/node_modules/typescript/lib/tsc.js:15748:24)
    at getEffectiveDecoratorFirstArgumentType (/usr/local/lib/node_modules/typescript/lib/tsc.js:21923:24)
    at getEffectiveDecoratorArgumentType (/usr/local/lib/node_modules/typescript/lib/tsc.js:22001:24)
    at getEffectiveArgumentType (/usr/local/lib/node_modules/typescript/lib/tsc.js:22014:24)
    at inferTypeArguments (/usr/local/lib/node_modules/typescript/lib/tsc.js:21795:35)
    at chooseOverload (/usr/local/lib/node_modules/typescript/lib/tsc.js:22147:33)
    at resolveCall (/usr/local/lib/node_modules/typescript/lib/tsc.js:22073:26)
```

### PR DESCRIPTION
**Pull request by yuit**
Fix #10193

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -337,8 +337,18 @@ namespace ts {
                         ? Diagnostics.Cannot_redeclare_block_scoped_variable_0
                         : Diagnostics.Duplicate_identifier_0;
 
+                    // If the current node has NodeFlags.Default (e.g. if the node is class declaration or function declaration)
+                    // and there is already another default export (i.e. symbol.declarations is not empty), we need to report such error
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: It took me a while to tie this together. How about

```text
// If the current node is a default export of some sort, then check if
// there are any other default exports that we need to error on.
// We'll know whether we have other default exports depending on if `symbol` already has a declaration list set.
```

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -337,8 +337,18 @@ namespace ts {
                         ? Diagnostics.Cannot_redeclare_block_scoped_variable_0
                         : Diagnostics.Duplicate_identifier_0;
 
+                    // If the current node has NodeFlags.Default (e.g. if the node is class declaration or function declaration)
+                    // and there is already another default export (i.e. symbol.declarations is not empty), we need to report such error
+                    if (isDefaultExport && symbol.declarations) {
+                        message = Diagnostics.A_module_cannot_have_multiple_default_exports;
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: Add a `Debug.assert(symbol.declarations.length > 0, "Expected at least one existing default declaration.")`

File: src/compiler/binder.ts [hunk 3]
```diff
@@ -337,8 +337,18 @@ namespace ts {
                         ? Diagnostics.Cannot_redeclare_block_scoped_variable_0
                         : Diagnostics.Duplicate_identifier_0;
 
+                    // If the current node has NodeFlags.Default (e.g. if the node is class declaration or function declaration)
+                    // and there is already another default export (i.e. symbol.declarations is not empty), we need to report such error
+                    if (isDefaultExport && symbol.declarations) {
+                        message = Diagnostics.A_module_cannot_have_multiple_default_exports;
+                    }
+
                     forEach(symbol.declarations, declaration => {
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: Do you need this anymore with the above check?

Comment (yuit) [inline src/compiler/binder.ts]: yep. Consider this 
```typescript
export default function Foo() { }
export default {
    "hi";1
}  // this has no NodeFlags.Default (as default is not modifier); we still need to loop through and check.  I should make comment
```

**Comment by yuit** [comment]
:bell:  @DanielRosenwasser

File: src/compiler/binder.ts [hunk 4]
```diff
@@ -337,11 +337,24 @@ namespace ts {
...
-                    });
+                    // If the current node is a default export of some sort, then check if
+                    // there are any other default exports that we need to error on.
+                    // We'll know whether we have other default exports depending on if `symbol` already has a declaration list set.
+                    if (isDefaultExport && symbol.declarations) {
+                        message = Diagnostics.A_module_cannot_have_multiple_default_exports;
+                    }
+                    else {
+                        // This is to properly report an error in the case "export default { }" is after export default of class declaration or function declaration.
+                        forEach(symbol.declarations, declaration => {
```
Comment (mhegazy) [inline src/compiler/binder.ts]: use for..of and beak on the first matching entry.

File: src/compiler/binder.ts [hunk 5]
```diff
@@ -349,11 +349,27 @@ namespace ts {
                             ? Diagnostics.Cannot_redeclare_block_scoped_variable_0
                             : Diagnostics.Duplicate_identifier_0;
 
-                        forEach(symbol.declarations, declaration => {
-                            if (declaration.flags & NodeFlags.Default) {
+                        if (symbol.declarations && symbol.declarations.length) {
```
Comment (mhegazy) [inline src/compiler/binder.ts]: Do we need to do this loop? what if we just do something like:

```ts
if (symbol.declarations && symbol.declarations.length && 
   (isDefaultExport || (node.kind === SyntaxKind.ExportAssignment &&  !(<ExportAssignment>node).isExportEquals)) {
    message = Diagnostics.A_module_cannot_have_multiple_default_exports;
}
```

**Comment by mhegazy** [comment]
@yuit what is the status of this PR?

**Comment by yuit** [comment]
@mhegazy  the pr slip my attention 😅 with all 2.0 bugs. gonna update now

**Pull request merged**